mithril-client/src/certificate_client/verify.rs

```diff
@@ -0,0 +1,495 @@
+            CertificateToVerify::ToDownload { hash } => {
+                self.retriever.get_certificate_details(&hash).await?
+            }
+        };
+
+        let previous_certificate = self
+            .internal_verifier
+            .verify_certificate(&certificate, &self.genesis_verification_key)
+            .await?;
+
+        #[cfg(feature = "unstable")]
+        if let Some(cache) = self.verifier_cache.as_ref() {
+            if !certificate.is_genesis() {
+                cache
+                    .store_validated_certificate(&certificate.hash, &certificate.previous_hash)
+                    .await?;
+            }
+        }
+
+        trace!(self.logger, "Certificate validated"; "hash" => &certificate.hash, "previous_hash" => &certificate.previous_hash);
+        self.feedback_sender
+            .send_event(MithrilEvent::CertificateValidated {
+                certificate_hash: certificate.hash,
+                certificate_chain_validation_id: certificate_chain_validation_id.to_string(),
+            })
+            .await;
+
+        Ok(previous_certificate.map(|cert| CertificateToVerify::Downloaded { certificate: cert }))
+    }
+}
+
+enum CertificateToVerify {
+    /// The certificate is already downloaded.
+    Downloaded { certificate: Certificate },
+    /// The certificate is not downloaded yet (since its parent was cached).
+    ToDownload { hash: String },
+}
+
+impl CertificateToVerify {
+    fn hash(&self) -> &str {
+        match self {
+            CertificateToVerify::Downloaded { certificate } => &certificate.hash,
+            CertificateToVerify::ToDownload { hash } => hash,
+        }
+    }
+}
+
+#[cfg_attr(target_family = "wasm", async_trait(?Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+impl CertificateVerifier for MithrilCertificateVerifier {
+    async fn verify_chain(&self, certificate: &MithrilCertificate) -> MithrilResult<()> {
+        // Todo: move most of this code in the `mithril_common` verifier by defining
+        // a new `verify_chain` method that take a callback called when a certificate is
+        // validated.
+        let certificate_chain_validation_id = MithrilEvent::new_certificate_chain_validation_id();
+        self.feedback_sender
+            .send_event(MithrilEvent::CertificateChainValidationStarted {
+                certificate_chain_validation_id: certificate_chain_validation_id.clone(),
+            })
+            .await;
+
+        // always validate the given certificate even if it was cached
+        let mut current_certificate = self
+            .verify_not_cached_certificate(
+                &certificate_chain_validation_id,
+                CertificateToVerify::Downloaded {
+                    certificate: certificate.clone().try_into()?,
+                },
+            )
+            .await?;
+
+        loop {
+            match current_certificate {
+                None => break,
+                Some(next) => {
```

Comment:
No need to ensure the absence of infinite loop ?

Reply:
We did not check it either before (and don't check it fully in the common verifier either as we only check that a certificate don't loop to itself), but this also bother me.

Reply:
Good point. Let's do that in another PR :+1: